src/hotspot/share/opto/addnode.cpp

```diff
@@ -852,6 +938,29 @@ Node* OrLNode::Ideal(PhaseGVN* phase, bool can_reshape) {
       return new RotateRightNode(in(1)->in(1), shift, TypeLong::LONG);
     }
   }
+  if (can_reshape && !phase->C->major_progress() && Matcher::match_rule_supported(Op_BitfieldInsertL)) {
+    Node *andl = NULL;
+    Node *dst = in(1);
+    Node *src = in(2);
+    int offset = 0;
+    if (src->Opcode() == Op_LShiftL && src->in(1)->Opcode() == Op_AndL && src->in(2)->is_Con()) {
+      andl   = src->in(1);
+      offset = src->in(2)->get_int();
+    } else if (src->Opcode() == Op_AndL) {
+      andl = src;
+    }
+    if (andl != NULL) {
+      Node* value = andl->in(1);
+      Node* mask  = andl->in(2);
+      if (mask->is_Con() && is_power_of_2(mask->get_long() + 1)) {
+        julong width = exact_log2_long(mask->get_long() + 1);
+        julong mask = ((1L << width) - 1) << offset;
+        if (width + offset <= 64 && ((value_range_maskL(phase, dst) & mask) == 0)) {
+          return new BitfieldInsertLNode(dst, value, phase->intcon(offset), phase->intcon(width));
+        }
+      }
+    }
+  }
```

Comment:
This needs reworking as above. I suggest you comment it by referring the reader back to the previous method.